crates/wasi-common/src/sys/stdio.rs

```diff
@@ -0,0 +1,107 @@
+use super::{fd, AsFile};
+use crate::handle::{Handle, HandleRights};
+use crate::sandboxed_tty_writer::SandboxedTTYWriter;
+use crate::wasi::types::{self, Filetype};
+use crate::wasi::{Errno, Result};
+use std::any::Any;
+use std::cell::Cell;
+use std::io::{self, Read, Write};
+
+pub(crate) trait StdioExt: Sized {
+    /// Create `Stdio` from `io::stdin`.
+    fn stdin() -> io::Result<Box<dyn Handle>>;
+    /// Create `Stdio` from `io::stdout`.
+    fn stdout() -> io::Result<Box<dyn Handle>>;
+    /// Create `Stdio` from `io::stderr`.
+    fn stderr() -> io::Result<Box<dyn Handle>>;
+}
+
+// The reason we have a separate Stdio type is to correctly facilitate redirects on Windows.
+// To elaborate further, in POSIX, we can get a stdio handle by opening a specific fd {0,1,2}.
+// On Windows however, we need to issue a syscall that's separate from standard Windows "open"
+// to get a console handle, and this is GetStdHandle. This is exactly what Rust does and what
+// is wrapped inside their Stdio object in the libstd. We wrap it here as well because of this
+// nuance on Windows:
+//
+//      The standard handles of a process may be redirected by a call to SetStdHandle, in which
+//      case GetStdHandle returns the redirected handle.
+//
+// The MSDN also says this however:
+//
+//      If the standard handles have been redirected, you can specify the CONIN$ value in a call
+//      to the CreateFile function to get a handle to a console's input buffer. Similarly, you
+//      can specify the CONOUT$ value to get a handle to a console's active screen buffer.
+//
+// TODO it might worth re-investigating the suitability of this type on Windows.
+#[derive(Debug, Clone)]
+#[allow(dead_code)]
+pub(crate) enum Stdio {
+    In { rights: Cell<HandleRights> },
+    Out { rights: Cell<HandleRights> },
+    Err { rights: Cell<HandleRights> },
+}
+
+impl Handle for Stdio {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn try_clone(&self) -> io::Result<Box<dyn Handle>> {
+        Ok(Box::new(self.clone()))
+    }
+    fn get_file_type(&self) -> Filetype {
+        Filetype::CharacterDevice
```

Comment:
Right, however, here, I think it's safe to assume that our internal `Stdio` type will always be a character device since it is meant to be a wrapper for Rust's stdio primitives. I envisioned that redirections should be handled by a separate entity in the codebase, namely, `OsOther` which is something like a catch-all for everything not covered with the specific handle types. Does that make sense? In fact, that's why in `ctx.rs` we map `std::fs::File` to `OsOther` as currently we only ever permit specifying custom fds/handles for stdio when builing the context.

Reply:
The logic for flushing Rust's stdout buffer is in the `Stdout` implementation here, and we need that for files too. I expect we're always going to want to use `Stdout` for stdout, even when it's attached to a file, in order to ensure that we always stay in sync with Rust's stdout.

Reply:
Hmm, I'm a little bit confused here. We cannot create a `std::io::Stdout` instance from a raw file descriptor, so I'm not sure we can ever enforce that we always use `sys::stdio::Stdout` for stdout. I think we could probably think of extracting `PipedStdout` or something from `OsOther` that would have `RawFd`/`RawHandle` attached to it.

Reply:
Oh, wait, I think I get it now. If the OS redirects the stdio handles, Rust's `Stdout` will use the redirected `RawFd` underneath, and this one doesn't necessarily have to be a character device. Is that what you meant?

Reply:
Fixed in 0945267. @sunfishcode if you could have a look and lemme know if that's what you had in mind!

Reply:
Yes, that's right.